some tests and readme

sarimabbas opened the pull request:
added tests to test_utils_lib, begun test_gripper, created read me, modified CMakeLists for tests

sarimabbas committed:
some tests and readme

alecive commented:
@sarimabbas nice pic! In general, it is a good practice to do smaller commits that are self-contained (eg one for the readme, one for the unit tests), and to push often, so that I don't have to review everything altogether.

Btw for now it's okay, I'll wait for travis and codacy and then I'll do my review.

sarimabbas commented:
Oh ok, will do next time!

alecive commented on baxter_collaboration_lib/CMakeLists.txt:
```diff
@@ -138,12 +138,19 @@ if(CATKIN_ENABLE_TESTING)
   add_rostest_gtest(test_robot_interface test/test_robot_interface.test
                                          test/test_robot_interface.cpp)
   target_link_libraries(test_robot_interface  robot_interface ${catkin_LIBRARIES})
+
```
Add comment as you did with `Robot interface tests` and "Robot util test"

alecive commented on baxter_collaboration_lib/test/README.md:
```diff
@@ -0,0 +1,49 @@
+## Testing
+
+#### Using [Catkin Tools](https://catkin-tools.readthedocs.io/en/latest/)
```
Remove one level of heading (also below)

alecive commented on baxter_collaboration_lib/test/README.md:
```diff
@@ -0,0 +1,49 @@
+	- [ ] ros_thread_obj.cpp
+
+* /baxter_collaboration_lib/src/robot_interface
+	- [x] robot_interface.cpp
```
`robotInterface` tests still need to be done.

alecive commented on baxter_collaboration_lib/test/test_gripper.cpp:
```diff
@@ -0,0 +1,38 @@
+	EXPECT_EQ(false, gr.is_enabled());
+	EXPECT_EQ(false, gr.is_calibrated());
+	EXPECT_EQ(false, gr.is_ready_to_grip());
+	EXPECT_EQ(false, gr.has_error());
```
This is not necessarily your fault, but prefer `EXPECT_FALSE` to `EXPECT_EQ(false, `. If that appeared in `test_robot_interface`, please change that as well.

alecive commented on baxter_collaboration_lib/test/test_gripper.cpp:
```diff
@@ -0,0 +1,38 @@
+	EXPECT_EQ(false, gr.has_error());
+
+	EXPECT_EQ(true, gr.is_sucking());
+	EXPECT_EQ(true, gr.is_gripping());
```
Simliarly, use `EXPECT_TRUE` here.

alecive commented on baxter_collaboration_lib/test/test_gripper.cpp:
```diff
@@ -0,0 +1,38 @@
+	EXPECT_EQ("left", gr.getGripperLimb());
+
+	//EXPECT_EQ(true, gr.gripObject());
+	//EXPECT_EQ(true, gr.releaseObject());
```
Delete these comments. These should be tested in the simulator test.

alecive commented on baxter_collaboration_lib/test/test_utils_lib.cpp:
```diff
@@ -89,6 +89,62 @@ TEST(UtilsLib, State)
+    testPose.orientation.x = 15; testPose.orientation.y = 20; testPose.orientation.z = 25; 
+    testPose.orientation.w = 30;
+    EXPECT_EQ("{position:{x: 0, y: 5, z: 10}, orientation:{x: 15, y: 20, z: 25, w: 30}}", toString(testPose));
+}
```
Good job here! But try also to think about edge cases, eg negative numbers, zeros, things that might break the code.

sarimabbas committed:
Updated testing CMakeLists and README

sarimabbas committed:
replaced EXPECT_EQ statements with EXPECT_BOOL statements for boolean results

sarimabbas commented:
Have reviewed everything except the edge cases for test_utils_lib, will think about them in detail today!

alecive commented on baxter_collaboration_lib/test/test_utils_lib.cpp:
```diff
@@ -89,6 +89,62 @@ TEST(UtilsLib, State)
 
 }
 
+// sarim's tests
```
Remove this comment, it's not useful :)

alecive commented on baxter_collaboration_lib/test/README.md:
````diff
@@ -0,0 +1,49 @@
+	```
+* If seeing unchanging results, `catkin build` may be required.
+
+### Checklist
````
I would probably move the checklist out of the `README`. Open an issue and place it there instead.

sarimabbas committed:
removed stray comment from test_utils_lib and removed checklist from README

sarimabbas committed:
updated edge case tests for utils and slightly modified README

added required dependency moveit to README, added edge-case tests to test_utils_lib, updated toString functions in utils.cpp and utils.h to handle negative zero floating point values

alecive commented on baxter_collaboration_lib/include/robot_utils/utils.h:
```diff
@@ -107,6 +107,16 @@ void setPosition(geometry_msgs::Pose& pose, float x, float y, float z);
+ *
+ * return N/A
+ */
+void subStringReplace(std::string& target, std::string replaceThis, std::string replaceWith);
```
What is this function for?

alecive replied:
I'm not saying that you should not have done it, but please justify :)

sarimabbas replied:
Sure. I wanted to preserve the code that was already present that used stringstream. But the str() method with stringstream would not convert something like -0.0 to 0 (instead it would give -0). So for just the toString(double) situation, the fix was easy enough, but for the toString(doubleVector) and toString(Pose) situations, there was the possibility of multiple -0s being present in the final string. So I thought I'd create a new function in utils that I could use for both situations.

Within the toString functions, a vector like [-0.0, 1, 2, 3, -0] now becomes [0, 1, 2, 3, 0]. Same for a Pose. I tried to think of potential dangers of such a function, but in my use, I have only targeted "-0," and "-0}" substrings for replacement.

alecive replied:
What if you change `toString(vector)` into a for loop that cycles through the elements and does individual `toString(double)`? Overall, this is a very minor function, whose behavior is not critical since it is only used for printouts. I would try to move on with this as fast as possible and start testing more important things.

sarimabbas replied:
You're right, this was a cleaner approach which I have implemented :)

alecive commented on baxter_collaboration_lib/test/README.md:
```diff
@@ -0,0 +1,30 @@
+4. Check the results of above: `catkin_test_results build/"$REPO_NAME"_lib || (cat build/"$REPO_NAME"_lib/test_results/"$REPO_NAME"_lib/*.xml ; false)`
+
+Troubleshooting:
+* If build fails, make sure prerequisites are installed: [`aruco_ros`](https://github.com/ScazLab/aruco_ros), `nlopt`, [`trac_ik`](https://bitbucket.org/alecive/trac_ik), [`baxter_common`](http://sdk.rethinkrobotics.com/wiki/Workstation_Setup#adeedda5360641914fe9c5d681c30026) and [`moveit`](http://moveit.ros.org/install/).
```
Why`moveit`?

sarimabbas replied:
When Andrew and I were trying to get him set up, he was unable to build baxter_collaboration without moveit. On the other hand, his trac_ik package was not building, so perhaps if his trac_ik builds, moveit may not be required. So I can defer this till tomorrow?

alecive replied:
Yes. Moveit is not needed in our software, let's take a look at that together.

sarimabbas committed:
Dealt with edge case alternatively, removed moveit from README

alecive committed:
[test] Cleanup , unified naming conventions, ready to merge after checks pass

alecive commented:
Ok feel free to merge, delete the branch and work on a new thing in a new branch.

sarimabbas commented:
Great! How would i do the merge? The button is greyed out on my end.

alecive commented:
Ops sorry now it should be working.

sarimabbas commented:
![screenshot from 2017-05-23 12 57 40](https://cloud.githubusercontent.com/assets/3000809/26366002/703ca6e0-3fb7-11e7-9a9b-47d8b62637d2.png)

Sorry but even after your approval it is greyed out
